CodingStandard.md

```diff
@@ -0,0 +1,1425 @@
+the typeid operator requires justification.
+
+**Friend**
+
+Using the friend keyword requires justification.
+
+**Protected**
+
+Using the protected keyword requires justification.
+
+**Initialization Lists**
+
+Constructor initializer lists must have the same order as the member
+data in the class definition.
+
+**Construction**
+
+Referencing objects within a constructor that are not member variables
+or arguments to that constructor requires justification showing that these
+objects have been constructed and initialized prior to the execution of the
+constructor.
+
+**Macros**
+
+Use of C preprocessor function-macros requires justification. Use of
+inline functions are preferred.
+
+Software Design Standards
+================================================================================
+
+**Software Design Constraints**
+
+- Global data must be documented in the design.
+- Global data must be minimized.
+- Bi-directional dependencies between software elements must be justified.
+- Exceptions and assertions used for control flow must be justified. Exception:
+  a timer exception used for task scheduling.
+- The use of self-modifying software is prohibited.
+
+**Software Design Guidelines**
+
+- The architecture should support one or more levels of abstraction with clearly
+  defined layers.
+- The architecture and design detail should be modular.
+- The design should reduce software complexity.
+- The design should promote reuse.
+- The design should facilitate extensibility.
+- The design should enhance maintainability.
+- The design should result in deterministic software with correct and
+  predictable behavior.
+- Justification will be required when considering software constructs or schemes
+  such as scheduling, interrupt and event-driven architectures, dynamic tasking,
+  re-entry, global data and exception handling.
+- The design should have high cohesion within software elements.
+- The design should have loose coupling between software elements.
+- The design should include well-defined interface descriptions between software
+  elements.
+- The design should include a hardware abstraction layer (HAL) that provides a
+  device-independent interface to hardware services.
+
+**Object-Oriented Software Design Guidelines**
+
+- Where appropriate, make use of proven design patterns. If a design pattern
+  adds complexity, make sure the benefit outweighs the added complexity.
+- Use of interfaces as a means to decrease coupling is recommended; allow
+  clients to depend on interfaces to software elements, not on their
+  implementations.
+- Composition/aggregation is recommended instead of implementation inheritance.
+- Use of inheritance solely to provision for unlikely future product
+  improvements should be avoided. Abstracting an interface and separating it
+  from an implementation as reasonable accommodation for change may be
+  appropriate.
+- Avoid allocating too many responsibilities to a single class.
+
```

Comment:
we should add a clause about doing our best to adhere to ROS constructs, and frameworks.